### PR TITLE
Setup Azure Pipelines

### PR DESCRIPTION
**Pull request by gnzlbg**
cc @alexcrichton this needs enabling Azure Pipelines for this repo

### CODE REVIEW

**Comment by gnzlbg** [comment]
cc @pietroalbini

**Comment by alexcrichton** [comment]
It's just sort of how it works but https://github.com/rust-lang/libc/pull/1377 is the branch auto-created by azure pipelines, and it should be configured now. It may or may not require actually using that particular branch, I'm not sure?

**Comment by alexcrichton** [comment]
The badge is:

[Build Status](https://dev.azure.com/rust-lang/libc/_build/latest?definitionId=11&branchName=master)

**Comment by gnzlbg** [comment]
@alexcrichton I see an azure job here that appears to properly show errors in the CI script :)

**Comment by alexcrichton** [comment]
👍 

Feel free to cc me if any further configuration is needed.

**Comment by gnzlbg** [comment]
So this is all green except for `aarch64-unknown-linux-android`, which appears to be missing for some reason under azure pipelines, but AFAICT it is there under travis. cc @alexcrichton @mati865 any idea what might be going on ?

**Comment by mati865** [comment]
@gnzlbg it should be `aarch64-linux-android`.

**Comment by gnzlbg** [comment]
Ok, so this is ready. There are two build jobs failing, Semver Linux&OSX, but these are expected to fail since we did some breaking changes. I've tried to use `continueOnError: true` to allow them to fail, without much luck, so I've temporarily disable them so that we can land this, and will re-enable them after we do a new release. If anybody knows what the Azure alternative for Travis `allow_failure`s is please let me know.

I think we need to land the PR to rust-central-station for bors to start using azure pipelines for this repo.

**Comment by mati865** [comment]
Yeah, it's not possible right now https://github.com/microsoft/azure-pipelines-tasks/issues/9302

**Comment by gnzlbg** [comment]
@mati865 I've tried to deploy your mingw appveyor linking fix without much luck - it also appears that the incorrect rustc version is being picked (1.35.0 as opposed to nightly). Any ideas ?

**Comment by mati865** [comment]
If toolchain is installed by rustup then `C:\Program Files (x86)\Rust\lib\rustlib\%TARGET%\lib` won't be valid location. Instead it should be something like `%USERPROFILE%\.rustup\toolchains\nightly-%TARGET%\lib\rustlib\%TARGET%\lib`.

I don't know Windows cmd well but with bash I'd use this location:
```
`rustc --print sysroot`/lib/rustlib/$TARGET/lib
```
It won't work for Rust installed via installer but should be fine for stable/beta/nightly installed by rustup.

**Comment by mati865** [comment]
```
D:\a\1\s>xcopy /Y "C:\msys64\mingw64\x86_64-w64-mingw32\lib\crt2.o" "C:\Program Files (x86)\Rust\lib\rustlib\x86_64-pc-windows-gnu\lib" 
File not found - crt2.o
0 File(s) copied
```
Logs from previous builds are not available, they could have location of MSYS2. Maybe it's not installed or installed to different directory.

File: ci/azure-install-rust.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      set -ex
+      gcc -print-search-dirs
+      find "C:\ProgramData\Chocolatey" -name "crt2*"
+      find "C:\ProgramData\Chocolatey" -name "dllcrt2*"
+      find "C:\ProgramData\Chocolatey" -name "libmsvcrt*"
+    condition: eq( variables['Agent.OS'], 'Windows_NT' )	
+    displayName: Find GCC libraries (windows)
+  - script: |	
+      @echo on	
+      if defined MSYS_BITS for %%I in (crt2.o dllcrt2.o libmsvcrt.a) do xcopy /Y "C:\ProgramData\Chocolatey\lib\mingw\tools\install\mingw64\x86_64-w64-mingw32\lib\%%I" "%USERPROFILE%\.rustup\toolchains\%TOOLCHAIN%-%TARGET%\lib\rustlib\%TARGET%\lib"
```
Comment (mati865) [inline ci/azure-install-rust.yml]: Failed with path: `C:\Users\VssAdministrator\.rustup\toolchains\-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\lib`
`%TOOLCHAIN%` is undefined :man_shrugging:

**Comment by mati865** [comment]
`if %ARCH% EQU i686 choco install mingw --x86 --force` part worked fine.

It's just copying from wrong directory to wrong directory:
```
D:\a\1\s>xcopy /Y "C:\ProgramData\Chocolatey\lib\mingw\tools\install\mingw64\x86_64-w64-mingw32\lib\crt2.o" "C:\Program Files\Rust\.rustup\toolchains\-i686-pc-windows-gnu\lib\rustlib\i686-pc-windows-gnu\lib\crt2.o" 
File not found - crt2.o
0 File(s) copied
```
For i686 that would be `C:\ProgramData\Chocolatey\lib\mingw\tools\install\mingw32\i686-w64-mingw32\lib\crt2.o`.
`C:\Program Files\Rust\.rustup\toolchains\-i686-pc-windows-gnu\lib\rustlib\i686-pc-windows-gnu\lib` is just totally wrong.
For rustup toolchains it should be `C:\Users\VssAdministrator\.rustup\toolchains\`**`nightly`**`-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\lib`. https://github.com/rust-lang/libc/pull/1376#discussion_r301555317

**Comment by gnzlbg** [comment]
So this is ready and green. We now spawn the jobs hierarchically, so it shouldn't fill the queue either one a single PR is being tested. 

When multiple PRs are on flight at the same time, Azure does not appear to have an option to limit the number of concurrent jobs per repository.

**Comment by mati865** [comment]
Just in case you want to see bash version of mingw script: https://dev.azure.com/mati865/_git/libc?path=%2Fci%2Fazure-install-rust.yml&version=GCcc8997972d5be0cf92a0f348970f2ab005ee6c46&line=49&lineStyle=plain&lineEnd=57&lineStartColumn=1&lineEndColumn=37
It takes path from `rustc` so there should be no problem if rustup is installed in weird place.

Passed tests:
https://dev.azure.com/mati865/libc/_build/results?buildId=20&view=results

**Comment by gnzlbg** [comment]
@pietroalbini this is timing out for some reason.

**Comment by pietroalbini** [comment]
Oh, I figured out what the problem is. You're running Azure Pipelines only on the `auto` and `try` branches, but bors is configured to [use `auto-libc` instead of `auto`](https://github.com/rust-lang/rust-central-station/blob/96e2562585ebb361a945b58d68230e81455cf025/homu.toml.template#L136-L137). If you replace the trigger with `["auto-libc", "try"]` everything should work.

**Comment by gnzlbg** [comment]
This is blocked on https://github.com/rust-lang/rust/issues/63115

**Comment by gnzlbg** [comment]
@rust-lang/infra any idea on why bors is continuously timing out ?

**Comment by pietroalbini** [comment]
Ok, bors successfully recognized Azure Pipelines :tada:

**Comment by gnzlbg** [comment]
This timeout appears to be correct. Azure Pipelines started executing some jobs, but at some point it hanged and stopped starting some of them. Cirrus-CI also appears to have hanged, with some jobs running for an unreasonable amount of time. 

I've cancelled the Azure jobs but cannot cancel the Cirrus one, so I'll wait till the queue there is clean and try again.

**Comment by mati865** [comment]
Something is wrong with Azure:
```
401 - Uh oh, you’re locked out.
Your account, <hidden> is not authorized to view this page. Make sure the URL is correct and your account has access. View a list of your organizations here. You may request access by contacting the admin.
08.08.2019 10:18:35 (UTC)
c0e79bdd-e865-496b-95bc-1dc982002779
```

**Comment by gnzlbg** [comment]
@mati865 

I think the homu azure job is still running, but it is only running one concurrent task, probably because the i686-pc-windows-gnu task (https://dev.azure.com/rust-lang2/libc/_build/results?buildId=65) failed with:

```
D:\a\1\s>if "i686" == "i686" choco install mingw --x86 --force 
Chocolatey v0.10.15
Installing the following packages:
mingw
By installing you accept licenses for the packages.
mingw v8.1.0 already installed. Forcing reinstall of version '8.1.0'. 
 Please use upgrade if you meant to upgrade to a new version.
mingw not installed. An error occurred during installation:
 The remote server returned an error: (503) Server Unavailable. Service Unavailable
This is try 1/3. Retrying after 300 milliseconds.
 Error converted to warning:
 Could not find a part of the path 'C:\ProgramData\chocolatey\lib\mingw\.chocolateyPending'.
This is try 2/3. Retrying after 400 milliseconds.
 Error converted to warning:
 Could not find a part of the path 'C:\ProgramData\chocolatey\lib\mingw\.chocolateyPending'.
Maximum tries of 3 reached. Throwing error.

Chocolatey installed 0/0 packages. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
Could not find a part of the path 'C:\ProgramData\chocolatey\lib\mingw\.chocolateyPending'.
##[error]Cmd.exe exited with code '1'.
```

**Comment by mati865** [comment]
I cannot access Azure at all: 
obraz

**Comment by gnzlbg** [comment]
@mati865 can you try again?

**Comment by mati865** [comment]
Thanks, I can access it now.

**Comment by gnzlbg** [comment]
So Azure succeeded and bors properly registered it: 🎉 .

Cirrus-CI is having issues, so I've canceled the builds for now. They will let us know when these issues are resolved.

**Comment by gnzlbg** [comment]
The Cirrus-CI issue should be resolved now.

@bors: retry

**Pull request merged**